main.cpp

```diff
@@ -19,13 +19,28 @@ using namespace clang::tooling;
 
 class CastCallBack : public MatchFinder::MatchCallback {
 public:
-    CastCallBack(Rewriter& rewriter) {
+    CastCallBack(Rewriter& _rewriter):rewriter(_rewriter) {
         // Your code goes here
     };
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+        auto csce = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+	if((csce!= nullptr)&&(!csce->getExprLoc().isMacroID())){
+	auto  csce_gs = csce->getSubExprAsWritten()->IgnoreImpCasts();

```

Comment:
Please describe why you are using here `getSubExprAsWritten()` and `IgnoreImpCasts()`.
What would be the difference with the use of `getSubExprAsWritten()->IgnoreImpCasts()`, `getSubExpr()->IgnoreImpCasts()` and `getSubExprAsWritten()`?

Reply:
If we add this code in run function:
`llvm::outs()<<"getSubExpr()  "<<E->getSubExpr()->getStmtClassName()<<"\n";`
`llvm::outs() <<"getSubExpr()->ignoreImpCasts()  "<< E->getSubExpr()->IgnoreImpCasts()->getStmtClassName()<<"\n";`
`llvm::outs() << "getSubExprAsWritten()  " <<E->getSubExprAsWritten()->getStmtClassName()<<"\n";`
`llvm::outs() << "getSubExprAsWritten()->ignoreImpCasts()  "<<E->getSubExprAsWritten()->ignoreImpCasts()->getStmtClassName()<<"\n";`
We can see the following output:
getSubExpr()  InplicitCastExpr
getSubExpr()->ignoreImpCasts()  DeclRefExpr
getSubExprAsWritten()  DeclRefExpr
getSubExprAsWritten()->ignoreImpCasts()  DeclRefExpr
The use of getSubExprAsWritten()->IgnoreImpCasts() makes no sence. We can replace this code part with getSubExpr()->ignoreImpCasts() or getSubExprAsWritten(). If we use getSubExpr(), we can't get the expected result.
Based on documentation, we can see the difference between these functions :
getSubExpr() retrieves the first cast subexpression.
getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
ignoreImpCasts() skips past any implicit casts which might surround this expression until reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.